### PR TITLE
Game/AI: Implement ContactDistance for ACTION_MOVE_TO_POS

### PR DESCRIPTION
**Pull request by Lumber-Jack-2**
[//]: # (***************************)
[//]: # (** FILL IN THIS TEMPLATE **)
[//]: # (***************************)

**Changes proposed:**

-  Add ContactDistance for ACTION_MOVE_TO_POS
-  This pr should make the script for https://github.com/TrinityCore/TrinityCore/pull/18702 more simple.
-  As you can see in https://github.com/TrinityCore/TrinityCore/pull/18702, the Peons should walk to the closest Wood Pile. However if we use ACTION_MOVE_TO_POS with targettype TARGET_CLOSEST_GAMEOBJECT, the peon will walk directly ON TOP of the gameobject. Blizzlike he should walk IN FRONT of the pile with a contactdistance from around 1 till 2 yards. With this PR we can add contactdistance to the ACTION_MOVE_TO_POS if a worldobject (gameobject, player, npc..) is used a targettype. 
-  This fix will not change the behaviour for TARGET_POSITION (intended)

Some screen for this:

GitHub Logo

**Target branch(es):** 3.3.5/master

- [X] 3.3.5

**Issues addressed:** Closes #  (insert issue tracker number)
- make script for https://github.com/TrinityCore/TrinityCore/pull/18702 more simple

**Tests performed:** (Does it build, tested in-game, etc.)
 yes and yes

**Known issues and TODO list:** (add/remove lines as needed)

- [x] One tiny problem. I hope the coredevs know the reason for it. If we use ACTION_MOVE_TO_POS with contactdistance = 0, the unit should actually walk INSIDE the target. But currently it alsways walks only 1 yard in front of the target. Dunno what causes that issue

I tested the script with this SAI (only customscript)

```sql
-- William Saldean SAI
SET @ENTRY := 33996;
UPDATE `creature_template` SET `AIName`="SmartAI" WHERE `entry`=@ENTRY;
DELETE FROM `smart_scripts` WHERE `entryorguid`=@ENTRY AND `source_type`=0;
INSERT INTO `smart_scripts` (`entryorguid`,`source_type`,`id`,`link`,`event_type`,`event_phase_mask`,`event_chance`,`event_flags`,`event_param1`,`event_param2`,`event_param3`,`event_param4`,`action_type`,`action_param1`,`action_param2`,`action_param3`,`action_param4`,`action_param5`,`action_param6`,`target_type`,`target_param1`,`target_param2`,`target_param3`,`target_x`,`target_y`,`target_z`,`target_o`,`comment`) VALUES
(@ENTRY,0,0,0,1,0,100,1,6000,6000,0,0,69,1,0,0,0,0,0,19,2118,100,0,0,0,0,0,"William Saldean - Out of Combat - Move To Closest Creature 'Abigail Shiel' (No Repeat)"),
(@ENTRY,0,1,0,34,0,100,0,0,1,0,0,51,0,0,0,0,0,0,1,0,0,0,0,0,0,0,"William Saldean - On Reached Point 1 - Kill Target");
```

[//]: # (**DEPRECATION NOTICE** Instead of creating new PRs to the (old) 6.x branch, make sure the target branch is **master** instead of 6.x.)
[//]: # (**NOTE** Enable the setting "[√] Allow edits from maintainers." when creating your pull request (if not enabled already).)
[//]: # (**NOTE** If this PR __only__ contains SQL files, open a new issue instead and post or link the SQL there.)
[//]: # (**SUGGESTION** When adding new SQL files, name them 9999_99_99_99_db_name.sql to reduce the chance of possible merge conflicts.)

### CODE REVIEW

**Comment by DDuarte** [comment]
Is it possible to make this configurable with a parameter somewhere in the SAI row? I do not have any example atm but it is very likely that someone will need the current behavior as well.

**Comment by Rushor** [comment]
if contactdistance is 0 it should walk on top, if contactdistance >0 it should calculate the distance, an additional param to enable/disable param contactdistance would blow sai up to much :(

**Comment by DDuarte** [comment]
Oh, if that already works, perfect! contactdistance = 0 is more than enough to signal that intention :)

**Comment by Rushor** [comment]
well, that's how it should be, but GetContactPoint currently always uses the objectsize to calculate the distance. I need some kind of function that calculates the distance without objectsize. i tried 

```sql
me->GetMotionMaster()->MovePoint(e.action.MoveToPos.pointId, target->GetNearPosition(e.action.MoveToPos.ContactDistance, target->GetAngle(me)));
```

doesn't pay attention to the position z :(

**Comment by Rochet2** [comment]
@Rushor if you look how GetContactPoint is coded you can see it uses
```cpp
GetNearPoint(obj, x, y, z, obj->GetObjectSize(), distance2d, GetAngle(obj));
```

youc an edit the object size to 0 probably. and use that function call in your code yourself

**Comment by Rushor** [comment]
ah yes, will try

```c++
float x, y, z;
                x = target->GetPositionX();
                y = target->GetPositionY();
                z = target->GetPositionZ();
                me->GetNearPoint(me, x, y, z, 0, e.action.MoveToPos.ContactDistance, 0);
                me->GetMotionMaster()->MovePoint(e.action.MoveToPos.pointId, x, y, z, e.action.MoveToPos.disablePathfinding == 0);
```

now, thanks :)

**Comment by Rushor** [comment]
alright i used this now:

```c++
                float x, y, z;
                x = target->GetPositionX();
                y = target->GetPositionY();
                z = target->GetPositionZ();
                target->GetNearPoint(me, x, y, z, 0, e.action.MoveToPos.ContactDistance, target->GetAngle(me));
                me->GetMotionMaster()->MovePoint(e.action.MoveToPos.pointId, x, y, z, e.action.MoveToPos.disablePathfinding == 0);

```

but the creature will still walk in front of the target if ContactDistance is 0 :(

GitHub Logo

**Comment by Shauren** [comment]
No, you cannot have object size 0

File: src/server/game/AI/SmartScripts/SmartScript.cpp [hunk 1]
```diff
@@ -1522,7 +1522,15 @@ void SmartScript::ProcessAction(SmartScriptHolder& e, Unit* unit, uint32 var0, u
                 me->GetMotionMaster()->MovePoint(e.action.MoveToPos.pointId, dest.x, dest.y, dest.z, e.action.MoveToPos.disablePathfinding == 0);
             }
             else
-                me->GetMotionMaster()->MovePoint(e.action.MoveToPos.pointId, target->GetPositionX(), target->GetPositionY(), target->GetPositionZ(), e.action.MoveToPos.disablePathfinding == 0);
+            {
+                float x, y, z;
+                x = target->GetPositionX();
+                y = target->GetPositionY();
+                z = target->GetPositionZ();
```
Comment (boxa) [inline src/server/game/AI/SmartScripts/SmartScript.cpp]: one call?
target->GetPosition(x, y, z);

Comment (Rushor) [inline src/server/game/AI/SmartScripts/SmartScript.cpp]: thx

**Comment by Rushor** [comment]
alright since nearly every distancecheck uses the contactdistance, i had to use: MovePosition()

```c++
                Position Pos = target->GetPosition();
                Pos.m_positionZ = target->GetPositionZ();
                me->MovePosition(Pos, e.action.MoveToPos.ContactDistance, me->GetAngle(target));
                me->GetMotionMaster()->MovePoint(e.action.MoveToPos.pointId, Pos, e.action.MoveToPos.disablePathfinding == 0);
```

This works fine now.

**Comment by Rushor** [comment]
well the creature will now ignore the objectsize of ContactDistance is 0 and will walk directly IN the target.

However if ContactDistance > 0 the angle where the creature runs to is broken.

Distance: 0 = http://i.imgur.com/HkCuPym.jpg
Distance: 5 = http://i.imgur.com/bRJSZlj.jpg

**Comment by Rochet2** [comment]
Did you try with target->GetAngle(me)?
Also with the current code it probably makes the npc move farther from the target if the npc is already within the contact distance. Is it supposed to be so that he moves away or that contact distance is the area where the npc does no longer need to move?

**Comment by Rushor** [comment]
yeah tried all kind of variations for the angle, it's always the same issue.

**Comment by Rochet2** [comment]
did you try to call the MovePosition through me? because it is likely me moving and not target.

**Comment by Rushor** [comment]
as i said, i tried every variation:
target->MovePosition(Pos, e.action.MoveToPos.ContactDistance, me->GetAngle(target));
target->MovePosition(Pos, e.action.MoveToPos.ContactDistance, target->GetAngle(me));

me->MovePosition(Pos, e.action.MoveToPos.ContactDistance, me->GetAngle(target));
me->MovePosition(Pos, e.action.MoveToPos.ContactDistance, target->GetAngle(me));

the angle is always wrong :D

**Comment by Rochet2** [comment]
GetAngle only returns (x,y) to (x,y) relative angles. Maybe what you need is an angle that also takes into account the current rotation of the npc.

Actually if you look at MovePosition implementation you can see that it does angle += GetOrientation();
You should use me->GetRelativeAngle(target) (or the other way around)

**Comment by Rushor** [comment]
tried:

me->MovePosition(Pos, e.action.MoveToPos.ContactDistance, target->GetRelativeAngle(me));
and 
me->MovePosition(Pos, e.action.MoveToPos.ContactDistance, me->GetRelativeAngle(target));

same issue as before :/

**Comment by Rushor** [comment]
alright i made it like this now:

> Is it possible to make this configurable with a parameter somewhere in the SAI row? I do not have any example atm but it is very likely that someone will need the current behavior as well (e.g move to right on top of the closest invisible trigger).

Now  it will work like this:
If the contactdistance is set to sth higher than 0, it will calculate the distance to the target, else it will all work like before (runnign directly on top of the target)

tested it ingame and it works fine :)

```sql
-- William Saldean SAI
SET @ENTRY := 33996;
UPDATE `creature_template` SET `AIName`="SmartAI" WHERE `entry`=@ENTRY;
DELETE FROM `smart_scripts` WHERE `entryorguid`=@ENTRY AND `source_type`=0;
INSERT INTO `smart_scripts` (`entryorguid`,`source_type`,`id`,`link`,`event_type`,`event_phase_mask`,`event_chance`,`event_flags`,`event_param1`,`event_param2`,`event_param3`,`event_param4`,`action_type`,`action_param1`,`action_param2`,`action_param3`,`action_param4`,`action_param5`,`action_param6`,`target_type`,`target_param1`,`target_param2`,`target_param3`,`target_x`,`target_y`,`target_z`,`target_o`,`comment`) VALUES
(@ENTRY,0,0,0,1,0,100,1,6000,6000,0,0,69,1,0,0,0,0,0,19,2118,100,0,0,0,0,0,"William Saldean - Out of Combat - Move To Closest Creature 'Abigail Shiel' (No Repeat)"),
(@ENTRY,0,1,0,34,0,100,0,0,1,0,0,51,0,0,0,0,0,0,1,0,0,0,0,0,0,0,"William Saldean - On Reached Point 1 - Kill Target");

and

-- William Saldean SAI
SET @ENTRY := 33996;
UPDATE `creature_template` SET `AIName`="SmartAI" WHERE `entry`=@ENTRY;
DELETE FROM `smart_scripts` WHERE `entryorguid`=@ENTRY AND `source_type`=0;
INSERT INTO `smart_scripts` (`entryorguid`,`source_type`,`id`,`link`,`event_type`,`event_phase_mask`,`event_chance`,`event_flags`,`event_param1`,`event_param2`,`event_param3`,`event_param4`,`action_type`,`action_param1`,`action_param2`,`action_param3`,`action_param4`,`action_param5`,`action_param6`,`target_type`,`target_param1`,`target_param2`,`target_param3`,`target_x`,`target_y`,`target_z`,`target_o`,`comment`) VALUES
(@ENTRY,0,0,0,1,0,100,1,6000,6000,0,0,69,1,0,0,5,0,0,19,2118,100,0,0,0,0,0,"William Saldean - Out of Combat - Move To Closest Creature 'Abigail Shiel' (5 yards dist.)(No Repeat)"),
(@ENTRY,0,1,0,34,0,100,0,0,1,0,0,51,0,0,0,0,0,0,1,0,0,0,0,0,0,0,"William Saldean - On Reached Point 1 - Kill Target");

thanks to @Riztazz and @DDuarte for the idea :)

**Comment by Rushor** [comment]
I rechcked this again now and everything works fine.
If the contactdistance is set to sth higher than 0, it will calculate the distance to the target, else it will all work like before (running directly on top of the target)

Anything else i need to add here before merge?

**Comment by Rushor** [comment]
the objectt that is used for the quest: 175783: TreeWayPoint (DND) (note for me)

and btw: merge this please.

**Comment by offl** [comment]
Also this should be used in #18444.

**Comment by Rushor** [comment]
yep ty :)

**Comment by Rushor** [comment]
pls merge this :(

**Comment by Rushor** [comment]
anyone? :)

**Comment by Rushor** [comment]
> anyone? :)

**Comment by Keader** [comment]
If no one has objections, i will merge it in 24h

**Pull request merged**